src/sentry/api/endpoints/organization_recent_searches.py

```diff
@@ -43,3 +68,30 @@ def get(self, request, organization):
         ).order_by('-last_seen')[:limit])
 
         return Response(serialize(recent_searches, request.user))
+
+    def post(self, request, organization):
+        serializer = RecentSearchSerializer(data=request.DATA)
+
+        if serializer.is_valid():
+            result = serializer.object
+
+            lock_key = 'recent_query:{}:{}:{}'.format(
+                six.text_type(request.user.id),
+                six.text_type(organization.id),
+                md5_text(six.text_type(result['type']) + result['query']).hexdigest(),
+            )
+            lock = locks.get(lock_key, duration=5)
+            with lock.acquire():
```

Comment:
Could use a retry lock here as well, I guess that'd prevent it just hard erroring if the lock can't be acquired...

Reply:
Do we often lock in endpoints? Locking up the web server could enable a denial of service if someone sends more concurrent requests than we have webserver threads. Doing this in celery would let you handle the locks more gracefully.

Reply:
I see them used in a fair few places, one example is `set_commits` which is called from a few different release detail endpoints.

That said, I don't mind moving this to a task. Do you know whether I need to add a `queue`, or is it fine to leave as defaults?

Reply:
You don’t need to make this an async task. Just to do two queries. 

Speaking of, why isn’t `remove_excess_recent_searches` just a single query?

I frankly don’t think we really need a lock here at all. The race condition isn’t that big of a deal if you do the delete in a single query instead of an O(n).

Reply:
Oh, just old Disqus habits. We usually tried to perform db operations individually to ensure that Django signals are fired, but seems like that's not a big deal here, so I'll change it to be a single query

Reply:
Using a single query is better than async task or a lock :clap: